Common URL Template path handling for REST requests (server and client)

ebullient opened the pull request:
Fixes #16952
Fixes #15231
Fixes #15044

Provide a common way to remember the parameterized path of a REST endpoint.

Both Micrometer and OpenTelemetry use parameterized values to reduce the cardinality of telemetry data, e.g. using `/hello/{name}` instead of `/hello/gary`. 

RESTEasy classic and RESTEasy Reactive manage creation/registration of templated paths very differently and two extensions need to be able to consume the parameterized version of the path. 

* Resteasy classic server requests: An additional interceptor annotation is added that stores the partly-resolved templated value from the original `@Path` annotation (taking a configured Rest ApplicationPath value into account). An associated interceptor binding looks up that annotation to retrieve the templated path, and set it on the common vertx context where consumers (both Micrometer and OpenTelemetry) can find it.

* Resteasy classic client requests:

ebullient commented on extensions/micrometer/deployment/pom.xml:
```diff
@@ -113,7 +119,7 @@
                 <configuration>
                     <systemPropertyVariables>
-                        <quarkus.log.level>INFO</quarkus.log.level>
+                        <quarkus.log.level>DEBUG</quarkus.log.level>
```
We need to put this back to INFO

ebullient commented on extensions/micrometer/deployment/src/test/resources/test-logging.properties:
```diff
@@ -1,4 +1,5 @@
+quarkus.log.category."io.quarkus.resteasy.observability".level=DEBUG
 quarkus.log.category."io.quarkus.bootstrap".level=INFO
-#quarkus.log.category."io.quarkus.arc".level=DEBUG
+quarkus.log.category."io.quarkus.arc".level=DEBUG
```
revert this change, too

ebullient commented on extensions/opentelemetry/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/vertx/VertxTracingAdapter.java:
```diff
@@ -1,15 +1,7 @@
-import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.HTTP_TARGET;
-import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.HTTP_URL;
-import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.HTTP_USER_AGENT;
+import static io.opentelemetry.semconv.trace.attributes.SemanticAttributes.*;
```
Was this intentional? Or is the IDE being too helpful

kenfinnigan replied:
Yeah, IDE

ebullient commented on extensions/resteasy-observability/deployment/src/main/java/io/quarkus/resteasy/observability/deployment/RestPathAnnotationProcessor.java:
```diff
@@ -0,0 +1,118 @@
+            builder.setRemovable();
+        }
+
+        return builder.build();
```
Rather than fuss with removable or not, just return null. Don't even create the beans.

Also, consider passing metricsCapability and capabilities into a readable method ( "hasConsumer" or something) that can be used in findRestPaths as well.

gsmet commented:
There's no detail in the description so it's hard to guess what this is for?

Until now, we didn't have specific extension for observability but we enabled it if the ad hoc extension was around and a config property enabled. What is the rationale for creating a specific extension here?

kenfinnigan commented:
@gsmet I can add a description, but the idea is we need a way for REST services to capture the parameterized version of a URL path for an endpoint as it's needed for setting on metrics and traces.

Not using the parameterized version results in high cardinality metrics and traces because we get a separate version for each parameter passed to the endpoint

kenfinnigan reviewed (dismissed):
I can approve, but seems odd given I contributed as well ;-)

geoand commented:
I can't really say I fully understand what's going on here, but IIUC you are adding a CDI interceptor everywhere there is a `@Path` annotation?
Have you benchmarked how much of a slowdown this incurs, if any?
Is the point of doing that is to stuff something in the Vertx context, right? If so, in RESTEasy Reactive there is definitely a way to do that without paying the cost of the the CDI interceptor.

Furthermore, is the new obversability module meant to be a hard dependency? Can't RESTEasy Classic and RESTEasy Reactive extensions just depend on an SPI module for the build items?

kenfinnigan commented:
> I can't really say I fully understand what's going on here, but IIUC you are adding a CDI interceptor everywhere there is a `@Path` annotation?

Yes, but only when Micrometer or OpenTelemetry are also present, otherwise it is not included.

> Have you benchmarked how much of a slowdown this incurs, if any?

Not at present, but the interceptor only reads an annotation and sets a value in the Vertx context, so it should be fairly light.

> Is the point of doing that is to stuff something in the Vertx context, right? If so, in RESTEasy Reactive there is definitely a way to do that without paying the cost of the the CDI interceptor.

Yes, that's right.

Do you have pointers to examples?

The reason a CDI Interceptor was chosen is we needed something that works for both RESTEasy Classic and RESTEasy Reactive. Maybe having the common annotation transformer is all that needs to be common, and we can have specific ways to retrieve the annotation and add it to the context.

> 
> Furthermore, is the new obversability module meant to be a hard dependency? Can't RESTEasy Classic and RESTEasy Reactive extensions just depend on an SPI module for the build items?

Yes, it's meant to be a hard dependency. There originally was an SPI with just the build item, but then it became too complicated as to how to include the actual deployment dependency given the logic of "if (RESTEasy Classic OR RESTEasy Reactive) AND (Micrometer OR OpenTelemetry)". It was decided the best option was to directly include in the REST extensions but have the processor do nothing if Micrometer or OpenTelemetry aren't present.

geoand commented:
> > I can't really say I fully understand what's going on here, but IIUC you are adding a CDI interceptor everywhere there is a `@Path` annotation?
> 
> Yes, but only when Micrometer or OpenTelemetry are also present, otherwise it is not included.
> 
> > Have you benchmarked how much of a slowdown this incurs, if any?
> 
> Not at present, but the interceptor only reads an annotation and sets a value in the Vertx context, so it should be fairly light.
> 

The readind the annotation part is what has me worried. We have tried very hard in RESTEasy Reactive to avoid reading annotations at runtime, so if this can be avoided, I would much prefer it.

> > Is the point of doing that is to stuff something in the Vertx context, right? If so, in RESTEasy Reactive there is definitely a way to do that without paying the cost of the the CDI interceptor.
> 
> Yes, that's right.
> 
> Do you have pointers to examples?
> 

Basically you would need to add a new `ServerRestHandler` and alter the handler chain with the `HandlerChainCustomizer` using the `BEFORE_METHOD_INVOKE` phase.
Any of the usages of `HandlerChainCustomizer` should give you an idea of what it does, but `org.jboss.resteasy.reactive.server.processor.scanning.CacheControlScanner` is probably a good place to start

> The reason a CDI Interceptor was chosen is we needed something that works for both RESTEasy Classic and RESTEasy Reactive. Maybe having the common annotation transformer is all that needs to be common, and we can have specific ways to retrieve the annotation and add it to the context.
> 
> > Furthermore, is the new obversability module meant to be a hard dependency? Can't RESTEasy Classic and RESTEasy Reactive extensions just depend on an SPI module for the build items?
> 
> Yes, it's meant to be a hard dependency. There originally was an SPI with just the build item, but then it became too complicated as to how to include the actual deployment dependency given the logic of "if (RESTEasy Classic OR RESTEasy Reactive) AND (Micrometer OR OpenTelemetry)". It was decided the best option was to directly include in the REST extensions but have the processor do nothing if Micrometer or OpenTelemetry aren't present.

OK

kenfinnigan commented:
> Reading the annotation part is what has me worried.
> We have tried very hard in RESTEasy Reactive to avoid reading annotations at runtime, so if this can be avoided, I would much prefer it.

Even when it's a straight read and then set without any computation associated with it?

What alternatives are available to do what is needed?

> 
> Basically you would need to add a new `ServerRestHandler` and alter the handler chain with the `HandlerChainCustomizer` using the `BEFORE_METHOD_INVOKE` phase.
> Any of the usages of `HandlerChainCustomizer` should give you an idea of what it does, but `org.jboss.resteasy.reactive.server.processor.scanning.CacheControlScanner` is probably a good place to start
> 

Thanks

geoand commented:
> > Reading the annotation part is what has me worried.
> > We have tried very hard in RESTEasy Reactive to avoid reading annotations at runtime, so if this can be avoided, I would much prefer it.
> 
> Even when it's a straight read and then set without any computation associated with it?
> 

Yes, our strategy has been to put this kind of thing in it's own `ServerRestHandler` that is only added to the processing pipeline if necessary 
> What alternatives are available to do what is needed?
> 
> > Basically you would need to add a new `ServerRestHandler` and alter the handler chain with the `HandlerChainCustomizer` using the `BEFORE_METHOD_INVOKE` phase.
> > Any of the usages of `HandlerChainCustomizer` should give you an idea of what it does, but `org.jboss.resteasy.reactive.server.processor.scanning.CacheControlScanner` is probably a good place to start
> 
> Thanks

👍🏼

kenfinnigan commented:
Where do we store the path value if not on an annotation? We need to store it somewhere between the annotation transformer in the build step and runtime.

geoand commented:
You could stuff it in the implementation of `ServerRestHandler` you use - much as `CacheControlHandler` does.

If you like, I can have a go at this on Monday.

kenfinnigan commented:
If you have some time, that would be appreciated

geoand commented:
Sure yeah, I'll take care of it Monday. I would do it tomorrow, but I have a talk to give so it's very unlikely I'll have the time :)

ebullient commented:
That doesn't resolve the issue, Georgios. Doing something special in Resteasy reactive doesn't fix/address the mess that is resteasy classic.
While this is "reading an annotation" .. it is doing it via a lookup in a table (I don't know if that is better or worse, but it is a lookaside because it's the special Arc transformed annotations, and not those that were originally on the method (discovered via reflection).

Part of what is nice about this solution is that it is comparatively simple. I've tried doing this 8 different ways, and this is the way that covers resteasy reactive AND resteasy classic (AND the clients in both cases).

geoand commented:
> That doesn't resolve the issue, Georgios. Doing something special in Resteasy reactive doesn't fix/address the mess that is resteasy classic.
> While this is "reading an annotation" .. it is doing it via a lookup in a table (I don't know if that is better or worse, but it is a lookaside because it's the special Arc transformed annotations, and not those that were originally on the method discovered via reflection).
> 
> Part of what is nice about this solution is that it is comparatively simple. I've tried doing this 8 different ways, and this is the way that covers resteasy reactive AND resteasy classic (AND the clients in both cases).

My objection is not about RESTEasy Classic. 
My focus is keeping RESTEasy Reactive as lean and mean as possible. What I propose would get around the reading the annotation at runtime part without sacrificing anything when it comes to functionality.
I'll push a commit to the PR on Monday hopefully that will affect RESTEasy Reactive only

ebullient commented:
That doesn't really help (and would it address clients, too?). We need all 4 addressed, consistently. Until/unless we deprecate/punt classic..

geoand commented:
The client can stay the way it is, the server is what I am worried about.

Too much of this sort of thing could easily accumulate down the road

ebullient commented:
That won't work, Georgios -- because we'd need this scan anyway (still), and then it becomes more complicated because we have to know not to do it only in x,y,z condition.. which makes it harder to maintain.

I get the tradeoff for performance here, but we need to knock it off with the one-offs. If you can handle both server & client consistently w/ RR (and set that attribute in the context) including the RestApplicationPath (but not the httpRoot), great. We can move this into resteasy classic. But if you're going to try something special for RR, please do BOTH server AND client.

geoand commented:
Well what you propose for the server doesn't work for me either :)

As I mentioned, we have put too much work into not reading data at runtime to start backtracking on this now

ebullient commented:
> Well what you propose for the server doesn't work for me either :)
> 
> As I mentioned, we have put too much work into not reading data at runtime to start backtracking on this now.

I get it, and that's fine... just please handle BOTH server and client. Then we're good.

geoand commented:
On that note, are there tests that cover these 4 cases?

ebullient commented:
> On that note, are there tests that cover these 4 cases?

I have some in micrometer that are disabled for now (because they have never worked).
Just have to change the result to show the templated path. (let me rummage, will update)

geoand commented:
👍🏼 

Those would be super helpful in making sure that everything works as expected

ebullient commented:
The deployment tests are using Resteasy classic right now.. 
the integration tests (which have RR) don't have the CLI stuff there yet. If you're waiting until Monday, I'll update this PR w/ those tests (Ken was focused on the tracing, so I was going to do the micrometer updates later) .. so I can have end to end server + client for RR and RC by Monday

geoand commented:
Sounds good to me. I was planning on picking it up Monday anyway

stuartwdouglas reviewed (changes requested):
In addition to the concerns @geoand raised I think there are a few other issues here. 

- The interceptor is not invoked on client requests. As these are synthetic beans they are not eligable for interception. 
- If the interceptor was invoked on client requests it would wipe out the server data. It is also possible to have multiple async client requests in flight at once, so save/restore is not an option, this data needs to be stored in an object that is scoped to the client request.
- Vertx.currentContext() should not be used, instead the CurrentVertxRequest should be injected into the interceptor, and `((HttpServerRequestInternal)request.getCurrent().request()).context()` should be used to get the context that is associated with the request, no matter the current thread local state. 

Basically on the server side the interceptor approach is fine for RESTEasy Classic, Resteasy Reactive should be changed to a handler based approach (which is easy). 

The client can't use this approach and needs to somehow extract the URL into the ClientRequestContext. This should be pretty easy for RR, but more of a problem for resteasy classic.

geoand commented:
I also have another question regarding the client:

What happens when a client request is made during an active server request?
It seems to me looking at the interceptor, that the Vertx context will contain the `UrlPathTemplate` value of the client and the server value will be lost.
How is that handled?

geoand commented:
I'll fix things on the RESTEasy Reactive side on Monday, but RESTEasy Classic is up you to address the issues based that Stuart brought up

stuartwdouglas commented:
As in mentioned in my previous comment the client is broken, it will overwrite the server version, and can't handle concurrent async requests.

ebullient commented:
As I am observing (putting the micrometer side of these tests together), the interceptor does not overwrite the server context because it uses a different context entirely (which the client filter running later doesn't see). So there is something else to fix there, I agree (and I've moved this PR back to draft while I sort that out).

I understand the desire to do something different for RR. The tests in extensions/micrometer/deployment cover resteasy classic (server and client) and integration-tests/micrometer-prometheus will cover resteasy reactive (server and client). The resteasy reactive tests (minimally) will be failing and ready for you to fix on Monday, @geoand =)

kenfinnigan commented:
Previously I'd moved away from `CurrentVertxRequest` usage because it didn't make the value available at the point the Tracing SPI is invoked at the end, but I think I might have been doing something slightly different than what you've proposed @stuartwdouglas so will try that approach again

ebullient commented:
> In addition to the concerns @geoand raised I think there are a few other issues here.
> 
>     * The interceptor is not invoked on client requests. As these are synthetic beans they are not eligable for interception.

Except that, in the case of resteasy classic, the interceptor is invoked. I can see it happening. The interceptor is not called for resteasy reactive, nor is the right context in the right place, so it just fails all over (using the same client filter for both, and maybe that needs to change, too?). 

>     * If the interceptor was invoked on client requests it would wipe out the server data. It is also possible to have multiple async client requests in flight at once, so save/restore is not an option, this data needs to be stored in an object that is scoped to the client request.

In resteasy classic, when the interceptor is invoked, it is using a different vertx context. The server value isn't overwritten. It still doesn't work, however, because the client filter sees the original server request context (it sees the server value), so it is still broken, but not due to a value overwrite.

>     * Vertx.currentContext() should not be used, instead the CurrentVertxRequest should be injected into the interceptor, and `((HttpServerRequestInternal)request.getCurrent().request()).context()` should be used to get the context that is associated with the request, no matter the current thread local state.

This should work for the server, but won't for the client. Again, the interceptor *does* get called for both. We'll have to know which is used for what. Ken -- if we're going to stick with this for resteasy classic, we could probably add a boolean indicating which path we're on. The transformer knows.. 

> Basically on the server side the interceptor approach is fine for RESTEasy Classic, Resteasy Reactive should be changed to a handler based approach (which is easy).

> The client can't use this approach and needs to somehow extract the URL into the ClientRequestContext. This should be pretty easy for RR, but more of a problem for resteasy classic.

I think we have to adjust the approach to look for something else to store it on, certainly... but as above, the interceptor is called (for resteasy classic client)

ebullient committed:
Micrometer and OpenTelemetry templated uri

 - Add tests for Micrometer and OpenTelemetry to verify server and client uri is parameterized
 - Add UrlTemplatePath handling for RESTEasy Reactive Server
 - Add UrlTemplatePath handling for Reactive REST Client
 - Move Opentelemetry listener registration to relevant extension
 - Use a DynamicFeature to set the UrlPathTemplate
 - Expand existing MethodScanner

Co-authored-by: Erin Schnabel <ebullient@redhat.com>
Co-authored-by: Ken Finnigan <ken@kenfinnigan.me>
Co-authored-by: Georgios Andrianakis <geoand@gmail.com>
Co-authored-by: Stuart Douglas <stuart.w.douglas@gmail.com>

geoand commented:
The RESTEasy Reactive server part is done and I am looking into the client part as well (which seems to be more involved). 
Not sure if I'll get it done today however since I have an Insights show and then need to step out to run some errands.

geoand commented:
I pushed a fix for the Reactive REST Client as well.

There is still one test failing:

` PrometheusMetricsRegistryTest.testClientRequests:87` but that fails with a `404` so I think there is something wrong with that test

ebullient commented:
> ` PrometheusMetricsRegistryTest.testClientRequests:87` but that fails with a `404` so I think there is something wrong with that test

Yep, there was. All fixed. Found another issue with the classic path, that's also been corrected. Let's see how CI does.

stuartwdouglas commented on extensions/resteasy-classic/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/QuarkusRestPathTemplateInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+            return context.proceed();
+        } finally {
+            if (annotation != null) {
+                ((HttpServerRequestInternal) request.getCurrent().request()).context().putLocal("UrlPathTemplate",
```
This needs to happen before the call, as async requests can complete before this method has actually returned. E.g. if you have the following code you will see the response being completed by the new thread before this call stack has returned:

```
    @GET
    @Path("gauge/{number}")
    public void checkListSize(@PathParam("number") long number, @Suspended AsyncResponse ar) throws Exception{
        if (number == 2 || number % 2 == 0) {
            // add even numbers to the list
            list.add(number);
        } else {
            // remove items from the list for odd numbers
            try {
                number = list.removeFirst();
            } catch (NoSuchElementException nse) {
                number = 0;
            }
        }
        long res = number;
        new Thread(new Runnable() {
            @Override
            public void run() {
                ar.resume(res);
            }
        }).start();
        Thread.sleep(1000);
    }
```

stuartwdouglas commented on extensions/resteasy-classic/resteasy/runtime/src/main/java/io/quarkus/resteasy/runtime/QuarkusRestPathTemplateInterceptor.java:
```diff
@@ -0,0 +1,49 @@
+        } finally {
+            if (annotation != null) {
+                ((HttpServerRequestInternal) request.getCurrent().request()).context().putLocal("UrlPathTemplate",
+                        annotation.value());
```
Note that this won't deal with subresources, but I don't know how big an issue this is. As sub resources are directly instantiated by the user code you can't guarantee that interceptors will be applied.

ebullient replied:
would a subresource have a path annotation?

stuartwdouglas replied:
Well the initial resource locator method will, however the resource that is returned likely won't be managed by CDI so won't be intercepted.

stuartwdouglas commented on extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ServerResourceMethodCustomizerBuildItem.java:
```diff
@@ -0,0 +1,60 @@
+     * Implementations will be invoked for each {@link ServerResourceMethod} discovered by Quarkus with the purpose of
+     * altering the ServerResourceMethod
+     */
+    public interface ServerResourceMethodCustomizer {
```
IMHO this sort of thing should be in the `indepdent-projects/resteasy-reactive` project, rather than the Quarkus extension.

stuartwdouglas replied:
@geoand I had a look at an alternate approach at https://github.com/quarkusio/quarkus/commit/7c1f1f0b505fe41285558b747b90e2eebce9b56b that just expands the existing MethodScanner stuff, WDYT?

geoand replied:
I was actually trying to force myself to do that but got to lazy 😸.

The reason I initially went down the road I did, was because I thought I needed the deployment path.
I only figured out that I didn't when I finished the implementation ☹️.
Feel free to push a commit including your change, as there is no good reason to go with my approach, all things considered

stuartwdouglas commented on independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/ResourceMethodCustomizer.java:
```diff
@@ -0,0 +1,27 @@
+import org.jboss.resteasy.reactive.common.model.ResourceClass;
+import org.jboss.resteasy.reactive.server.model.ServerResourceMethod;
+
+public interface ResourceMethodCustomizer {
```
Looks like this is not used, and the BuildItem version is used instead?

stuartwdouglas commented:
Test failure looks related, it looks like a META-INF/services entry is not registered in native mode.

kenfinnigan commented on extensions/resteasy-classic/rest-client/deployment/src/main/java/io/quarkus/restclient/deployment/RestClientProcessor.java:
```diff
@@ -269,6 +279,70 @@ void processInterfaces(
+
+            for (MethodInfo methodInfo : classInfo.methods()) {
+                if (!isDefault(methodInfo.flags())) {
+                    templates.put(interfaceName.toString() + "." + methodInfo.name(), constructPath(methodInfo));
```
@ebullient made me aware of that on the weekend, and I will be looking at it this morning.

Any tips on a good key to differentiate methods with the same name?

geoand replied:
The only reliable way is to use the types of the arguments in the key.

kenfinnigan replied:
And I now see you've already fixed it! Thanks!

geoand replied:
I didn't, but maybe Stuart did :)

kenfinnigan replied:
Yeah, Stuart did

geoand commented:
> I probably shouldn't approve as I was involved in it

@kenfinnigan we have all been involved in it, so that would apply to any of our reviews :)

ebullient reviewed (commented):
The changes that aren't mine LGTM! (but I can't approve as it is my PR)
